notebooks/linear_regression.livemd

```diff
@@ -0,0 +1,610 @@
+```
+
+```elixir
+# Shuffle data to make splitting more resonable
+{num_rows, _num_cols} = Explorer.DataFrame.shape(df)
+permutation_indices = Nx.iota({num_rows}) |> Nx.shuffle(axis: 0)
+
+y =
+  df[["median_house_value"]]
+  |> ToTensor.df_to_matrix()
+  |> Nx.take(permutation_indices)
+  |> Nx.squeeze()
+
+x =
+  df
+  |> Explorer.DataFrame.discard(["median_house_value"])
+  |> ToTensor.df_to_matrix()
+  |> Nx.take(permutation_indices)
+
+{x, y}
+```
+
+Since we don't have a stratified spilt of data implemented (to learn more see [Stratified Sampling](https://en.wikipedia.org/wiki/Stratified_sampling)), we shuffle data set and take advantage of [Law of large numbers](https://en.wikipedia.org/wiki/Law_of_large_numbers). Now, we will split the data into training and test sets.
+
+```elixir
+train_ratio = 0.8
+num_train = round(train_ratio * num_rows)
+
+{x_train, x_test} = {x[[0..(num_train - 1), 0..-1//1]], x[[num_train..(num_rows - 1), 0..-1//1]]}
+{y_train, y_test} = {y[[0..(num_train - 1)]], y[[num_train..(num_rows - 1)]]}
+```
+
+Before we compute the correlation matrix, we will check if we have NaNs (Not a Number) in the data set.
+
+```elixir
+y_nan_count = Nx.sum(Nx.is_nan(y))
+x_nan_count = Nx.sum(Nx.is_nan(x))
+{x_nan_count, y_nan_count}
+```
+
+Ups, we have some. Fortunately, for y, we don't have any NaNs. If we dig a little bit more, it turns out that NaNs are in <pre> bedrooms_per_rooms (1st row) </pre>  <pre> total_bedrooms (10th row) </pre>
+
+```elixir
+{bedrooms_per_rooms_idx, total_bedrooms_idx} = {0, 9}
+bedrooms_per_rooms_nan_count = Nx.sum(Nx.is_nan(x[[0..-1//1, bedrooms_per_rooms_idx]]))
+total_bedrooms_nan_count = Nx.sum(Nx.is_nan(x[[0..-1//1, total_bedrooms_idx]]))
+Nx.equal(x_nan_count, Nx.add(bedrooms_per_rooms_nan_count, total_bedrooms_nan_count))
+```
+
+For these two, we use `Scholar.Impute.SimpleImputer` with startegy set to median of values. Function `fit` learn the median of features and `transform` for trained model replace all NaNs with a given startegy. It is important that we perform imputation after splitting data because otherwise we will have a leakage of information from test data.
+
+```elixir
+x_train =
+  x_train
+  |> Scholar.Impute.SimpleImputer.fit(strategy: :median)
+  |> Scholar.Impute.SimpleImputer.transform(x_train)
+
+x_test =
+  x_test
+  |> Scholar.Impute.SimpleImputer.fit(strategy: :median)
+  |> Scholar.Impute.SimpleImputer.transform(x_test)
+```
+
+Eventually, we can compute the correlation matrix. We will use `Scholar.Covariance` to calculate the correlation matrix.
+
+```elixir
+defmodule Correlation do
+  import Nx.Defn
+
+  defn correlation_matrix_n(x, opts \\ []) do
+    variances =
+      if opts[:biased] do
+        Nx.variance(x, axes: [0])
+      else
+        Nx.variance(x, axes: [0], ddof: 1)
+      end
+
+    Scholar.Covariance.covariance_matrix(x, opts) /
+      Nx.sqrt(Nx.new_axis(variances, 1) * Nx.new_axis(Nx.transpose(variances), 0))
```

Comment:
Why do you need the transpose here?

Reply:
Won't `variances` be 1D and as such transpose be a noop?

Reply:
Yes, you are totally right. It is redundant.